var/spack/repos/builtin/packages/qmcpack/package.py

```diff
@@ -349,38 +355,35 @@ def install(self, spec, prefix):
     # QMCPACK 3.6.0 install directory structure changed, thus there
     # thus are two version of the setup_environment method
     @when('@:3.5.0')
-    def setup_environment(self, spack_env, run_env):
+    def setup_run_environment(self, env):
         """Set-up runtime environment for QMCPACK.
         Set PYTHONPATH for basic analysis scripts and for Nexus."""
-        run_env.prepend_path('PYTHONPATH', self.prefix.nexus)
+        env.prepend_path('PYTHONPATH', self.prefix.nexus)
 
     @when('@3.6.0:')
-    def setup_environment(self, spack_env, run_env):
+    def setup_run_environment(self, env):
         """Set-up runtime environment for QMCPACK.
         Set PYTHONPATH for basic analysis scripts and for Nexus. Binaries
         are in the  'prefix' directory instead of 'prefix.bin' which is
         not set by the default module environment"""
-        run_env.prepend_path('PATH', self.prefix)
-        run_env.prepend_path('PYTHONPATH', self.prefix)
+        env.prepend_path('PATH', self.prefix)
+        env.prepend_path('PYTHONPATH', self.prefix)
 
     @run_after('build')
     @on_package_attributes(run_tests=True)
-    def check(self):
+    def check_install(self):
```

Comment:
Why was this renamed? Actually, CMakePackage already runs `check` by default during test time, so if you change it back to `check` you should be able to remove the `run_after` and `on_package_attributes`.

Reply:
`check` did not seem to do anything anymore, so I changed it to `check_install` because I grepped through some packages and found an example similar to the one I had.

You are saying to get rid of the decorators and just do `check` and it will run after the build -- and if the build fails, it will not install. correct?

Reply:
Correct. You can test this with `spack install --test=root qmcpack` to make sure I'm right.

Reply:
Unfortunately, your suggested modifications are not working. `check` is doing nothing.

Reply:
I'm having trouble reproducing this. I tried messing with a simpler package with fewer dependencies. I tried building with and without overriding `check`, and with and without extending `CudaPackage`, but it always builds tests when I ask it to. What version of Python are you using to run Spack?

Reply:
Looks like its using `python3`. Did your test use CMake?

Reply:
I tried both Python 2 and 3 and couldn't reproduce the bug. Yes, I was using a `CMakePackage`. Can you upload your build log?

Reply:
OK, now I see what is happening  and I was really confused. `install --test-root` does run `check`, but its not the custom tests that I want. It seems to be running some default CMake unit tests. They run so fast that I repeatedly missed seeing them run. `check_install` is a custom method and runs a more aggressive tests that take a much longer time to run. It was only running when the conditions in the decorator were true. (the presence of `--run-tests`). I would like to be able to run both sets of tests, but I am not sure that its possible right now with Spack. I am ok with the way the PR is now, though I have to admit its kind of confusing:
1. `install --test=root`, runs some default unit tests
2. `--run-tests`, runs the custom check_install method

Especially when the output of `spack install --help` suggests that `--run-tests` is equivalent to `install --test=all`

Any advice? Like to get this PR request merged in sooner rather than later.

Reply:
`--test=root` and `--run-tests` should be equivalent, are you testing this on develop? If you want to run both `make check` and your custom unit tests, your current approach is fine, or you can add `make check` to your custom function and override the default `check`.

Reply:
I also tried combining the decorators with the custom `check` method and it successfully overrides the builtin `check` method when I use `--test=root`.

Reply:
> `--test=root` and `--run-tests` should be equivalent, are you testing this on develop? If you want to run both `make check` and your custom unit tests, your current approach is fine, or you can add `make check` to your custom function and override the default `check`.

yes

Reply:
Make sure you're using `--test=root` and not `--test-root`

Reply:
It does not seem to be working that way for me. If I rename the method to `check` instead of `check_install` and use `--test=root` or `--run-tests` the custom `check` (with the decorates present) is not run. 

Just merge 'AS IS.' If I can narrow down the problem with a simple reproducer, I will file a github issue.